### PR TITLE
Platform 5120: add /v1/professional/creators/async/contents/fetch API

### PR DESCRIPTION
**Pull request by vasu-iiq**
image
image
image

### CODE REVIEW

File: app/services/webhook_service.py [hunk 1]
```diff
@@ -86,6 +87,7 @@ async def process_webhook(webhook_request_data: WebhookRequestData):
         WebhookEvent.CONTENTS_FETCH_SUCCESS, WebhookEvent.CONTENTS_FETCH_FAILURE
     ] and Product.PUBLIC_CONTENT_SEARCH in settings.SUPPORTED_PRODUCTS:
         await process_contents_fetch_event(webhook_request_data=webhook_request_data)
+        await process_professional_contents_fetch_event(webhook_request_data=webhook_request_data)
```
Comment (anujith-s) [inline app/services/webhook_service.py]: Depending on WP ID, we need to decide which one of the two to trigger
Have a whitelist of WP IDs for professional work platforms, if the WP ID belongs to that then process professional, else social

Ex: PROFESSIONAL_WORK_PLATFORM_IDS = ['<ID_FOR_LI>']

File: app/services/webhook_service.py [hunk 2]
```diff
@@ -457,12 +461,32 @@ async def process_contents_fetch_event(webhook_request_data: WebhookRequestData)
...
     await send_events(webhook_event=webhook_request_data.event, data=async_contents)
 
 
+async def process_professional_contents_fetch_event(webhook_request_data: WebhookRequestData):
+    job_id = webhook_request_data.data.get("job_id")
+    work_platform = webhook_request_data.work_platform
+
+    professional_contents_fetch_event = ProfessionalContentsFetchEvent(
+        job_id=job_id,
+        work_platform=work_platform
```
Comment (anujith-s) [inline app/services/webhook_service.py]: why do you need work_platform here?

File: app/schemas/webhook_schemas.py [hunk 3]
```diff
@@ -83,3 +90,8 @@ class ContentsFetchEvent(BaseModel):
 
 class AudienceOverlapEvent(BaseModel):
     job_id: str
+
+
+class ProfessionalContentsFetchEvent(BaseModel):
+    job_id: str
+    work_platform: WorkPlatform
```
Comment (anujith-s) [inline app/schemas/webhook_schemas.py]: why is this needed?

**Review by shahaab-iiq — DISMISSED** [review]

File: app/api_v1/endpoints/professional_creators/api_professional_creator_processor.py [hunk 4]
```diff
@@ -12,3 +12,9 @@
 @api_router.post("/profiles/analytics", status_code=HTTPStatus.OK, response_model=Optional[Dict])
 async def profile_analytics(request: Request, request_body: Optional[object] = Body(default=None)) -> Optional[Dict]:
     return await creator_service.professional_profile_analytics(request_body=request_body, params=dict(request.query_params.items()))
+
+
+@api_router.post("/contents/fetch", status_code=HTTPStatus.OK, response_model=Optional[Dict])
+async def fetch_professional_contents(request: Request,
```
Comment (anujith-s) [inline app/api_v1/endpoints/professional_creators/api_professional_creator_processor.py]: make_professional_contents_fetch_request or post_professional_contents_fetch

File: app/events/base_event.py [hunk 5]
```diff
@@ -226,3 +226,15 @@ async def audience_overlap_success_event_handler(cls, data: Dict):
     @classmethod
     async def audience_overlap_failure_event_handler(cls, data: Dict):
         pass
+
+    @classmethod
+    async def professional_contents_fetch_request_event_handler(cls, data: Dict):
+        pass
+
+    @classmethod
+    async def professional_contents_fetch_request_success_handler(cls, data: Dict):
```
Comment (anujith-s) [inline app/events/base_event.py]: success event is not a request
```suggestion
    async def professional_contents_fetch_success_event_handler(cls, data: Dict):
```

File: app/events/db/db_table_field_mappings.yaml [hunk 6]
```diff
@@ -826,3 +826,83 @@ mappings:
...
             profiles_json: app.events.db.value_post_processors.cast_to_json
             ignored_profiles_json: app.events.db.value_post_processors.cast_to_json
+
+
+    professional_contents_fetch_data:
+        name: professional_contents_fetch_data
+        schema: iiq_schema
+        unique: professional_contents_fetch_request_iiq_id, platform_content_id
+        fields:
+            id: professional_contents_fetch_request_iiq_id
```
Comment (anujith-s) [inline app/events/db/db_table_field_mappings.yaml]: id could lead to confusion that, it is content id, but since it is actually job_id, we can rename it to that
```suggestion
            job_id: professional_contents_fetch_request_iiq_id
```

File: app/events/base_event.py [hunk 7]
```diff
@@ -226,3 +226,15 @@ async def audience_overlap_success_event_handler(cls, data: Dict):
     @classmethod
     async def audience_overlap_failure_event_handler(cls, data: Dict):
         pass
+
+    @classmethod
+    async def professional_contents_fetch_request_event_handler(cls, data: Dict):
```
Comment (anujith-s) [inline app/events/base_event.py]: request is not an event type
```suggestion
    async def professional_contents_fetch_request_handler(cls, data: Dict):
```

File: app/services/resource_service.py [hunk 8]
```diff
@@ -276,6 +276,13 @@ async def fetch_async_contents_by_id(id: str) -> Dict:
     return await fetch_all_responses_iteratively(url=url)
 
 
+@retry(attempts=3, delay=1, retry_exceptions=(TooManyRequestException,))
+async def fetch_async_professional_contents_by_id(id: str) -> Dict:
```
Comment (anujith-s) [inline app/services/resource_service.py:280]:
```suggestion
async def fetch_async_professional_contents_by_job_id(job_id: str) -> Dict:
```

Comment (anujith-s) [inline app/services/resource_service.py]:
```suggestion
async def fetch_async_professional_contents_by_job_id(job_id: str) -> Dict:
```

File: app/services/webhook_service.py [hunk 9]
```diff
@@ -7,12 +7,20 @@
...
 from app.services.resource_service import fetch_contents_by_ids, fetch_account_by_id, fetch_content_groups_by_ids, \
     fetch_profile_by_id, fetch_social_transactions_by_ids, fetch_social_payouts_by_ids, fetch_balances_by_ids, \
     fetch_activity_artists_by_ids, fetch_activity_contents_by_ids, fetch_profile_audience_by_account_id, \
     fetch_content_comments_by_content_id_account_id, fetch_commerce_transactions_by_ids, fetch_commerce_payouts_by_ids, \
-    fetch_publish_content_by_id, fetch_profile_analytics_by_id, fetch_async_contents_by_id, fetch_audience_overlap_by_id
+    fetch_publish_content_by_id, fetch_profile_analytics_by_id, fetch_async_contents_by_id, fetch_audience_overlap_by_id, \
+    fetch_async_professional_contents_by_id
+
+# whitelisted work_platform_ids for professional/contents
+PROFESSIONAL_WORK_PLATFORM_IDS = ["36410629-f907-43ba-aa0d-434ca9c0501a"]
```
Comment (anujith-s) [inline app/services/webhook_service.py]: * Move this to `app/utils/constants.py`
* Create a constant for `36410629-f907-43ba-aa0d-434ca9c0501a` this WP, then add that constant to the list
* Since IDs are not easily understood by humans, it requires an extra step to determine which work platform an ID belongs to

File: app/services/webhook_service.py [hunk 10]
```diff
@@ -7,12 +7,20 @@
...
+
+
+def is_professional_platform(work_platform_id: str) -> bool:
```
Comment (anujith-s) [inline app/services/webhook_service.py]: Move this to `app/utils/generic_utils.py`

File: app/services/webhook_service.py [hunk 11]
```diff
@@ -457,12 +474,28 @@ async def process_contents_fetch_event(webhook_request_data: WebhookRequestData)
...
+
+    professional_contents_fetch_event = ProfessionalContentsFetchEvent(job_id=job_id)
+
+    professional_contents_fetch_event_job_id = professional_contents_fetch_event.job_id
+
+    professional_contents: Dict = await fetch_async_professional_contents_by_id(id=professional_contents_fetch_event_job_id)
+
+    if not professional_contents:
+        logging.error(f"Professional-Contents do not exists with publish-id: {professional_contents_fetch_event_job_id}")
+        return
```
Comment (anujith-s) [inline app/services/webhook_service.py]: just simplifying things, nothing major
```suggestion
    event = ProfessionalContentsFetchEvent(**webhook_request_data.data)

    professional_contents: Dict = await fetch_async_professional_contents_by_id(id=event.job_id)

    if not professional_contents:
        logging.error(f"Professional-Contents do not exists with publish-id: {event.job_id}")
        return
```

File: app/events/db/db_table_field_mappings.yaml (hunk 6 above)
Comment (vasu-iiq) [reply app/events/db/db_table_field_mappings.yaml]: but this `id` is the column name that we are getting from response. Our table column name is `professional_contents_fetch_request_iiq_id` so we cannot change the `id` column

Comment (anujith-s) [reply app/events/db/db_table_field_mappings.yaml]: Isn't the id being populated here?

https://github.com/insightiq-ai/iiq-managed-service/blob/99e9957b2947ab23bb8101db43cf31f281ee3927/app/events/db/db_event_handler.py#L336-L340

File: app/services/resource_service.py (hunk 8 above)
Comment (anujith-s) [reply app/services/resource_service.py]: The argument also would change to `job_id`

**Review by anujith-s — APPROVED** [review]

**Review by shahaab-iiq — APPROVED** [review]
